src/neo/Consensus/ConsensusService.cs

```diff
@@ -185,8 +186,36 @@ private void InitializeConsensus(byte viewNumber)
             {
                 ChangeTimer(TimeSpan.FromMilliseconds(Blockchain.MillisecondsPerBlock << (viewNumber + 1)));
             }
+
+            // Try to speed up consensus if cached future payloads exists
+            if (hasFuturePayload)
+            {
+                TryToConsumeFuturePayload(context.FuturePreparationPayloads);
+                TryToConsumeFuturePayload(context.FutureCommitPayloads);
+                TryToConsumeFuturePayload(context.FutureChangeViewPayloads);
+                TryToConsumeFuturePayload(context.FutureRecoveryPayloads);
+            }
         }
 
+        private void TryToConsumeFuturePayload(ConsensusPayload[] payloadsArray)
+        {
+            for (int p = 0; p < payloadsArray.Length; p++)
+            {
+                var payload = payloadsArray[p];
+                if (payload != null)
+                {
+                    if (payload.BlockIndex < context.Block.Index)
+                    {
+                        payloadsArray[p] = null;
+                    }
+                    else if (payload.PrevHash == context.Block.PrevHash && payload.BlockIndex == context.Block.Index)
```

Comment:
@erikzhang, @shargon, talking to @vang1ong7ang he detected a possible attack here.
When we change the `consensus validators`, an old validated payload may be processed.

In this sense, we need to check the current `context.validators[payload.ValidatorIndex]` public key with the one provided on `payload.Witness`.

If you have an example of this code please send it here...aehauheua it may save some time.

Reply:
If you just send the payloads to `OnConsensusPayload()`, it will verify them, right?

Reply:
I think it will not, @erikzhang, the verification of the payload witness happens in a lower layer.

In these cases we are just simulating that the ConsensusService already pass such verification. I believe it is almost the same that we do on the Recovery.

Reply:
My fault, @erikzhang.

I forgot to call `ReverifyAndProcessPayload(ConsensusPayload payload)`.

Reply:
Done.

An optimization is still possible but can be done later.